src/devices/Ak8963/Ak8963.cs

```diff
@@ -0,0 +1,275 @@
+            magneto.Y = BinaryPrimitives.ReadInt16LittleEndian(rawData.Slice(2));
+            magneto.Z = BinaryPrimitives.ReadInt16LittleEndian(rawData.Slice(4));
+
+            if (OutputBitMode == OutputBitMode.Output16bit)
+            {
+                // From the documentation range is from 32760 which does represent 4912 µT
+                magneto *= 4912.0f / 32760.0f;
+            }
+            else
+            {
+                magneto *= 4912.0f / 8192.0f;
+            }
+            return magneto;
+
+        }
+
+        /// <summary>
+        /// Get the Magetometer data and wait for them to be ready
+        ///         +X
+        ///  \  |  /
+        ///   \ | /
+        ///    \|/
+        ///    /|\
+        ///   / | \
+        ///  /  |  \
+        ///    +Z   +Y
+        /// </summary>
+        public Vector3 Magnetometer => ReadMagnetometer(true);
+
+        /// <summary>
+        /// Get or set the device self test mode. 
+        /// If set to true, this creates a magnetic field
+        /// Once you read it, you will have the results of the self test
+        /// 14-bit output(BIT=“0”)  
+        ///          | HX[15:0]        | HY[15:0]        | HZ[15:0] 
+        /// Criteria | -50 =< HX =< 50 | -50 =< HY =< 50 | -800 =< HZ =< -200 
+        /// 16-bit output(BIT=“1”)  
+        ///          | HX[15:0]          | HY[15:0]          | HZ[15:0] 
+        /// Criteria | -200 =< HX =< 200 | -200 =< HY =< 200 | -3200 =< HZ =< -800 
+        /// </summary>
+        public bool MageneticFieldGeneratorEnabled
+        {
+            get => _selfTest;
+
+            set
+            {
+                byte mode = value ? (byte)0b01000_0000 : (byte)0b0000_0000;
+                WriteRegister(Register.ASTC, mode);
+                _selfTest = value;
+            }
+        }
+
+        /// <summary>
+        /// Select the measurement mode
+        /// </summary>
+        public MeasurementMode MeasurementMode
+        {
+            get => _measurementMode;
+
+            set
+            {
+                byte mode = (byte)((byte)value | ((byte)_outputBitMode << 4));
+                WriteRegister(Register.CNTL, mode);
+                _measurementMode = value;
+                // according to docuementation:
+                // After power-down mode is set, at least 100µs(Twat) is needed before setting another mode
+                Thread.Sleep(1);
+            }
+        }
+
+        /// <summary>
+        /// Select the output bit rate
+        /// </summary>
+        public OutputBitMode OutputBitMode
+        {
+            get => _outputBitMode;
+
+            set
+            {
+                byte mode = (byte)(((byte)value << 4) | (byte)_measurementMode);
+                WriteRegister(Register.CNTL, mode);
+                _outputBitMode = value;
+            }
+        }
+
+        private void WriteRegister(Register reg, byte data) => _ak8963Interface.WriteRegister(_i2cDevice, reg, data);
```

Comment:
These should just have the implementation of WriteRegister using I2c instead of forwarding the call to another class. This just makes the design more complicated than it needs to be.

Reply:
I have to pass a specific I2C device which does not really operate the same way. It's actually similar design as in the CharacterLcd: https://github.com/dotnet/iot/blob/master/src/devices/CharacterLcd/Hd44780.cs#L104 and it's for the same reason